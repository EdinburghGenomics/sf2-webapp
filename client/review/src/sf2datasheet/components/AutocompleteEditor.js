// @flow
import React, {PureComponent} from 'react';
import Autocomplete from 'react-autocomplete'

import * as R from 'ramda';
import {ENTER_KEY, TAB_KEY} from '../constants/keys';


// Flow type declarations ----

type OptionValueType = string;


type EventType = {
    which: number,
    persist: () => void
};


type AutocompleteEditorProps = {
    cell: Object,
    row: number,
    col: number,
    value: OptionValueType,
    onChange: () => void,
    onCommit: (OptionValueType, ?Object) => void,
    onRevert: () => void,
    onKeyDown: () => void,
    options: Array<string>,
    uuid: string
}


type AutocompleteEditorState = { e?: ?Object, value: string, selectedValue: string, lastPressedKey: number, selectionMade: boolean};


// React component ----

export default class AutocompleteEditor extends PureComponent<AutocompleteEditorProps, AutocompleteEditorState> {
    input: Autocomplete;


    constructor (props : AutocompleteEditorProps) {
        super(props);
        this.state = {
            value: '',
            selectedValue: '',
            lastPressedKey: -1,
            selectionMade: false
        };
        this.input = {focus: null, props: null};
    }


    commit = (value : string) => {

        const {onCommit, onRevert} = this.props;

        if (!value) {
            return onRevert()
        } else {
            const { e } = this.state;
            onCommit(value, e);
        }

    };


    handleSelect  = (value : string) => {

        // Logic to commit selected item via mouse click on menu option is handled here
        if(this.state.lastPressedKey !== ENTER_KEY && this.state.lastPressedKey !== TAB_KEY) {
            this.commit(value);
        } else {
            this.setState({
                selectedValue: value,
                selectionMade: true
            });
        }

    };

    getHighlightedItem = (value: string, options: Array<string>) : string => {
        const filteredOptions = R.filter(option => option.startsWith(value), options);
        return filteredOptions.length === 0 ? '' : filteredOptions[0];
    };


    handleKeyDown  = (e : EventType) : void =>  {

        this.setState({
            lastPressedKey: e.which
        });

    };

    componentDidMount() {

        if(this.input !== null && this.input !== undefined) {
            this.input.focus();
        }

    }

    // $FlowFixMe
    componentDidUpdate(prevProps, prevState) {

        // Logic to commit selected item via enter or tab key is handled here
        const highlightedItem = this.getHighlightedItem(this.state.value, this.props.options);

        if (
                (this.state.lastPressedKey === ENTER_KEY) ||
                (this.state.lastPressedKey === TAB_KEY && this.state.selectedValue === '' && highlightedItem === '')
        ) {
            this.commit(this.state.value);
        } else if (this.state.lastPressedKey === TAB_KEY && this.state.selectedValue !== '') {
            this.commit(this.state.selectedValue);
        }

    }


    render () {
        return (
            <span onKeyDown={this.handleKeyDown} style={{display: "flex", height: "100%"}}>
            <Autocomplete
                ref={el => this.input = el}
                items={this.props.options.map(val => {return {id: val, label: val}})}
                shouldItemRender={(item, value) => item.label.toLowerCase().indexOf(value.toLowerCase()) > -1}
                getItemValue={item => item.label}
                wrapperStyle={{ display: 'flex', width: "-webkit-fill-available" }}
                renderItem={(item, highlighted) =>
                    <div
                        key={item.id}
                        data-uuid={this.props.uuid}
                        style={{ backgroundColor: highlighted ? '#eee' : 'transparent'}}
                    >
                        {item.label}
                    </div>
                }
                renderInput={props => <input style={{ display: 'flex', width: "-webkit-fill-available" }} {...props} />}
                value={this.state.value}
                onChange={e => this.setState({ value: e.target.value })}
                onSelect={this.handleSelect}
                selectOnBlur={true}
                open={true}
                autoHighlight={true}
            /></span>
        )
    }
}
