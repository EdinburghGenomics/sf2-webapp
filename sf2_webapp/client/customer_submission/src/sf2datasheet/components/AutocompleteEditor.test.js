import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import AutocompleteEditor from './AutocompleteEditor';


// Basic page logic

it('renders AutocompleteEditor correctly', () => {

    const wrapper = mount(
        <AutocompleteEditor
            cell={{}}
            row={1}
            col={1}
            value='test'
            onChange={() => {}}
            onCommit={(test, e) => {}}
            onRevert={() => {}}
            onKeyDown={() => {}}
            options={['one', 'two']}
        />
    );

    expect(toJson(wrapper)).toMatchSnapshot();

});
