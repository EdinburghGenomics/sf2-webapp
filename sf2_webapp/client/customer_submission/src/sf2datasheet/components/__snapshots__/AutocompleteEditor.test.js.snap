// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders AutocompleteEditor correctly 1`] = `
<AutocompleteEditor
  cell={Object {}}
  col={1}
  onChange={[Function]}
  onCommit={[Function]}
  onKeyDown={[Function]}
  onRevert={[Function]}
  options={
    Array [
      "one",
      "two",
    ]
  }
  row={1}
  value="test"
>
  <span
    onKeyDown={[Function]}
    style={
      Object {
        "display": "flex",
        "height": "100%",
      }
    }
  >
    <Autocomplete
      autoHighlight={true}
      getItemValue={[Function]}
      inputProps={Object {}}
      isItemSelectable={[Function]}
      items={
        Array [
          Object {
            "id": "one",
            "label": "one",
          },
          Object {
            "id": "two",
            "label": "two",
          },
        ]
      }
      menuStyle={
        Object {
          "background": "rgba(255, 255, 255, 0.9)",
          "borderRadius": "3px",
          "boxShadow": "0 2px 12px rgba(0, 0, 0, 0.1)",
          "fontSize": "90%",
          "maxHeight": "50%",
          "overflow": "auto",
          "padding": "2px 0",
          "position": "fixed",
        }
      }
      onChange={[Function]}
      onMenuVisibilityChange={[Function]}
      onSelect={[Function]}
      open={true}
      renderInput={[Function]}
      renderItem={[Function]}
      renderMenu={[Function]}
      selectOnBlur={true}
      shouldItemRender={[Function]}
      value=""
      wrapperProps={Object {}}
      wrapperStyle={
        Object {
          "display": "flex",
          "width": "-webkit-fill-available",
        }
      }
    >
      <div
        style={
          Object {
            "display": "flex",
            "width": "-webkit-fill-available",
          }
        }
      >
        <input
          aria-autocomplete="list"
          aria-expanded={true}
          autoComplete="off"
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          role="combobox"
          style={
            Object {
              "display": "flex",
              "width": "-webkit-fill-available",
            }
          }
          value=""
        />
        <div
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onTouchStart={[Function]}
          style={
            Object {
              "background": "rgba(255, 255, 255, 0.9)",
              "borderRadius": "3px",
              "boxShadow": "0 2px 12px rgba(0, 0, 0, 0.1)",
              "fontSize": "90%",
              "left": 0,
              "maxHeight": "50%",
              "minWidth": 0,
              "overflow": "auto",
              "padding": "2px 0",
              "position": "fixed",
              "top": 0,
            }
          }
        >
          <div
            key="one"
            onClick={[Function]}
            onMouseEnter={[Function]}
            style={
              Object {
                "backgroundColor": "transparent",
              }
            }
          >
            one
          </div>
          <div
            key="two"
            onClick={[Function]}
            onMouseEnter={[Function]}
            style={
              Object {
                "backgroundColor": "transparent",
              }
            }
          >
            two
          </div>
        </div>
      </div>
    </Autocomplete>
  </span>
</AutocompleteEditor>
`;
